# DS-SF-36 | 03 | `pandas` | Assignment | Starter Code

## `pandas` and Bistro

You've just told one of your friend that you are taking a Data Science class.  (Yeah!)  Your friend is running a bistro, a small restaurant, serving moderately priced simple meals in a modest setting ([Wikipedia](https://en.wikipedia.org/wiki/Bistro)).  She collected over some period of time the following information of her patrons' visits.

| Variable's name | Its meaning |
|:---:|:---|
| `name` | Patron's first name |
| `gender` | Patron's gender |
| `is_smoker` | Whether the patron is smoking or not |
| `party` | Party's size |
| `check` | Check amount (\$) (after taxes but before tips) |
| `tip` | Tip (\$) that the patron added to the check |
| `day` | Week day of the visit |
| `time` | Rough time estimate of the visit |

In this assignment, we will be exploring this dataset using `pandas`.<sup>(*)</sup>

<sup>(*)</sup> this dataset was adapted from the `tips` dataset of the `seaborn` package (https://github.com/mwaskom/seaborn-data)

> ### Question 1.  Import `pandas` (as `pd`).

In [1]:
import os

import numpy as np

# TODO
import pandas as pd

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

In [5]:
!cd ../datasets

In [6]:
!dir

 Volume in drive C has no label.
 Volume Serial Number is 64EB-2D41

 Directory of C:\Users\Jmiller16\DS-SF-36\classes\03\notebooks

06/30/2017  04:53 PM    <DIR>          .
06/30/2017  04:53 PM    <DIR>          ..
06/30/2017  04:53 PM    <DIR>          .ipynb_checkpoints
06/30/2017  04:53 PM            10,729 assignment-03-pandas-starter-code-jessica.ipynb
06/28/2017  06:34 PM            11,672 assignment-03-pandas-starter-code.ipynb
06/28/2017  09:34 PM           234,565 codealong-03-pandas-jessica.ipynb
06/28/2017  06:34 PM            34,827 codealong-03-pandas-starter-code.ipynb
               4 File(s)        291,793 bytes
               3 Dir(s)  350,106,902,528 bytes free


> ### Question 2.  Read the `dataset-03-bistro.csv` dataset.

In [2]:
# TODO
df=pd.read_csv(os.path.join('..', 'datasets', 'dataset-03-bistro.csv'))

> ### Question 3.  What is the class of the `pandas` object storing the dataset?

In [7]:
# TODO
type(df)

pandas.core.frame.DataFrame

Answer: Panda DataFrame

> ### Question 4.  How many samples (i.e., rows) are in this dataset?

In [11]:
# TODO
df.shape[0]

244

Answer: 244

> ### Question 5.  How many variables (i.e., columns) are in this dataset?

In [13]:
# TODO
df.shape[1]

8

Answer: 8

> ### Question 6.  Print the name of each column in the dataset, one name per line.

In [20]:
# TODO
df.dtypes

day           object
time          object
name          object
gender        object
is_smoker       bool
party          int64
check        float64
tip          float64
dtype: object

> ### Question 7.  Print the first two rows of the dataset to the console.  What does the output look like?

In [23]:
# TODO
df.head(2)

day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66

Answer: TODO

> ### Question 8.  Extract the last 2 rows of the data frame and print them to the console.  What does the output look like?

In [24]:
# TODO
df.tail(2)

day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00

Answer: TODO

> ### Question 9.  Does the dataset contain any missing values?

In [26]:
# TODO
df.isnull().sum()

day          0
time         0
name         0
gender       0
is_smoker    0
party        0
check        0
tip          0
dtype: int64

Answer: Dataset has no null values

> ### Question 10.  What can you say about the `is_smoker` variable?  I.e., will it bring any insights when analyzing the dataset?  What do you want to do with it?  (and do it...)

In [27]:
# TODO
df.is_smoker.unique()

array([False], dtype=bool)

Answer: This variable is not useful in analyzing the dataset because it only has 1 value for the entire dataset.

> ### Question 11.  For which week days does the dataset has data for?

In [29]:
# TODO
df.day.unique()

array(['Sunday', 'Saturday', 'Thursday', 'Friday'], dtype=object)

Answer: Dataset has data for Thursday, Friday, Saturday and Sunday.

> ### Question 12.  How often was the bistro patronized for each week day?

(check `.value_counts()`; it could come in handy)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)

In [33]:
# TODO
df.day.value_counts()

Saturday    87
Sunday      76
Thursday    62
Friday      19
Name: day, dtype: int64

Answer: Bistro was patronized 62 times on Thursday, 19 times on Friday, 87 times on Saturday and 76 times on Sunday.

> ### Question 13.  How much tip did waiters collect for each week day?

In [68]:
# TODO
#df.groupby(['day']).sum().tip
df[['day','tip']].groupby('day').sum()

tip
day             
Friday     51.96
Saturday  260.40
Sunday    247.39
Thursday  171.83

Answer: waiters collected 171.83 on Thursday, 51.96 on Friday, 260.40 on Saturday and 247.39 on Sunday.

> ### Question 14.  What is the average tip per check (in absolute \$) for each week day?

In [69]:
# TODO
#df.groupby(['day']).mean().tip
df[['day','tip']].groupby('day').mean()

tip
day               
Friday    2.734737
Saturday  2.993103
Sunday    3.255132
Thursday  2.771452

Answer: The average tip is 2.77 on Thursday, 2.73 on Fridays, 2.99 on Saturday and 3.26 on Saturday.

> ### Question 15.  What is the average tip per check (as a percentage of the check) for each week day?

In [77]:
# TODO

df.groupby(['day']).sum().tip/df.groupby(['day']).sum().check


day
Friday      0.159445
Saturday    0.146424
Sunday      0.152038
Thursday    0.156732
dtype: float64

Answer: Thursday 16.13%, Friday 16.99%, Saturday 15.31%, Sunday 16.69%

In [78]:
df['tip_percentage']=df.tip/df.check
df[['day','tip_percentage']].groupby('day').mean()

tip_percentage
day                     
Friday          0.169913
Saturday        0.153152
Sunday          0.166897
Thursday        0.161276

> ### Question 16.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

(check `numpy.intersect1d()`; it could come in handy)

(https://docs.scipy.org/doc/numpy/reference/generated/numpy.intersect1d.html)

In [66]:
# TODO
import numpy as np

l1=df[df.gender=='Female'].name
l2=df[df.gender=='Male'].name
np.intersect1d(l1, l2)

array(['Casey'], dtype=object)

Answer: Casey

> ### Question 17.  If no patrons share the same name, how many unique patrons are in the dataset?

In [119]:
# TODO
len(df[df.gender=='Male'].name.unique())+len(df[df.gender=='Female'].name.unique())

182

Answer: 182

> ### Question 18.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [82]:
# TODO
df[df.name=='Alice'].count()

day               2
time              2
name              2
gender            2
is_smoker         2
party             2
check             2
tip               2
tip_percentage    2
dtype: int64

Answer: Kevin visited 4 times, Alice visited 2 times.

> ### Question 19.  Who are the top 3 female and male patrons?

In [113]:
# TODO
df_f=df[df.gender=='Female'].name.value_counts()
df_f.head(3)


Mary     4
Laura    3
Casey    3
Name: name, dtype: int64

In [112]:
df_m=df[df.gender=='Male'].name.value_counts()
df_m.head(3)

David    8
James    5
Casey    5
Name: name, dtype: int64

Answer: Top Females: Mary(4), Laura(3), Casey(3).  Top Males: David(8), James (5), Casey (5)

> ### Question 20.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html)

In [122]:
# TODO
(df.groupby(['gender','name']).tip.sum()/df.groupby(['gender','name']).check.sum()).sort_values(ascending=False)


gender  name    
Female  Maryann     0.416667
        Bailey      0.325733
Male    Daniel      0.313607
        Dennis      0.280535
        Zackary     0.266312
                      ...   
        Willie      0.066534
Female  Kimberly    0.059447
        Destiny     0.056797
        Mildred     0.056433
Male    Jeremy      0.035638
Length: 182, dtype: float64

In [124]:
df[df.name=='Maryann']

day    time     name  gender  is_smoker  party  check  tip  \
178  Sunday  Dinner  Maryann  Female      False      2    9.6  4.0   

     tip_percentage  
178        0.416667

In [125]:
df[df.name=='Jeremy']

day    time    name gender  is_smoker  party  check   tip  \
237  Saturday  Dinner  Jeremy   Male      False      2  32.83  1.17   

     tip_percentage  
237        0.035638

Answer: Maryann is the best tipper and Jeremy is the worst tipper.  They have both come once.